### PR TITLE
MAINT: blanket `noqa` vet

### PR DESCRIPTION
**Pull request by lucascolley**
#### Reference issue
N/A

#### What does this implement/fix?
- Removes all existences of blanket `noqa` by replacing with the specific errors
- Removes most unused `noqa`s (apart from those which are relevant for codes we may wish to check in the future)
- Configures the linter to reject blanket `noqa`s in the future

#### Additional information
Blanket `noqa`s can hide other violations and it only takes a few seconds to add in the right error code. Getting rid of them now will make any future work on the linter easier.

### CODE REVIEW

File: doc/source/conf.py [hunk 1]
```diff
@@ -212,7 +212,7 @@
 if 'dev' in version:
     html_theme_options["switcher"]["version_match"] = "development"
 
-if 'versionwarning' in tags:  # noqa
+if 'versionwarning' in tags:  # noqa: F821
```
Comment (lucascolley) [inline doc/source/conf.py:215]: I assume this is just going over my head but is `tags` actually defined here?

File: tools/lint.toml [hunk 2]
```diff
@@ -20,8 +20,7 @@ target-version = "py39"
 "scipy/stats/distributions.py" = ["F403", "F405"]
 "scipy/stats/mstats.py" = ["F403", "F401"]
 "scipy/stats/qmc.py" = ["F403"]
-"scipy/optimize/_linprog.py" = ["F401"]
-"scipy/optimize/_linprog_ip.py" = ["F401"]
+"scipy/stats/tests/data/_mvt.py" = ["E701", "E702"]
```
Comment (lucascolley) [inline tools/lint.toml:23]: I assume the multiple statements on one line are intentional here, @mdhaber ?

File: doc/source/conf.py (hunk 1 above)
Comment (j-bowhay) [reply doc/source/conf.py:215]: https://www.sphinx-doc.org/en/master/usage/configuration.html

File: scipy/stats/_levy_stable/__init__.py [hunk 3]
```diff
@@ -837,8 +837,8 @@ def _parameterization(self):
     def rvs(self, *args, **kwds):
         X1 = super().rvs(*args, **kwds)
 
-        discrete = kwds.pop("discrete", None)  # noqa
-        rndm = kwds.pop("random_state", None)  # noqa
+        _ = kwds.pop("discrete", None)
+        _ = kwds.pop("random_state", None)
```
Comment (j-bowhay) [inline scipy/stats/_levy_stable/__init__.py:841]:
```suggestion
        kwds.pop("discrete", None)
        kwds.pop("random_state", None)
```

File: tools/lint.toml (hunk 2 above)
Comment (mdhaber) [reply tools/lint.toml:23]: Yes. The intent was to translate a Matlab function as directly as possible.

File: benchmarks/benchmarks/cutest/dfovec.py [hunk 4]
```diff
@@ -312,7 +312,7 @@ def dfovec(m, n, x, nprob):
             tmp2 = np.exp(-x[5] * (temp - x[8]) ** 2)
             tmp3 = np.exp(-x[6] * (temp - x[9]) ** 2)
             tmp4 = np.exp(-x[7] * (temp - x[10]) ** 2)
-            fvec[i] = y5[i] - (x[0] * tmp1 + x[1] * tmp2 + x[2] * tmp3 + x[3] * tmp4)  # noqa
+            fvec[i] = y5[i] - (x[0] * tmp1 + x[1] * tmp2 + x[2] * tmp3 + x[3] * tmp4)
```
Comment (mdhaber) [inline benchmarks/benchmarks/cutest/dfovec.py:315]: Do you know what this was for?

File: scipy/_lib/_docscrape.py [hunk 5]
```diff
@@ -542,12 +542,12 @@ def __str__(self, func_role=''):
         return '\n'.join(out)
 
 
-def indent(str, indent=4):  # noqa
+def indent(str, indent=4):
```
Comment (mdhaber) [inline scipy/_lib/_docscrape.py:545]: Or this one?

File: scipy/optimize/_minpack_py.py [hunk 6]
```diff
@@ -862,7 +862,7 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
     suggesting that the optimal values of these parameters are ambiguous and
     that only one of these parameters is needed in the model.
 
-    """  # noqa
+    """
```
Comment (mdhaber) [inline scipy/optimize/_minpack_py.py:865]: Or this one?

**Review by mdhaber — APPROVED** [review]
I'm curious about the `noqa`s that were removed without replacement. You explained the bad variable names one, but I'm wondering about other rules that we are no longer checking.

File: benchmarks/benchmarks/cutest/dfovec.py (hunk 4 above)
Comment (lucascolley) [reply benchmarks/benchmarks/cutest/dfovec.py:315]: no

File: scipy/_lib/_docscrape.py (hunk 5 above)
Comment (lucascolley) [reply scipy/_lib/_docscrape.py:545]: using `str` as a variable name. If we ever wanted to enable the check for that, we would probably want this to be changed anyway.

Comment (mdhaber) [reply scipy/_lib/_docscrape.py:545]: Yeah OK.

File: benchmarks/benchmarks/cutest/dfovec.py (hunk 4 above)
Comment (lucascolley) [reply benchmarks/benchmarks/cutest/dfovec.py:315]: actually, probably PEP8 line length. It's <88 though so can be removed.

File: scipy/optimize/_minpack_py.py (hunk 6 above)
Comment (lucascolley) [reply scipy/optimize/_minpack_py.py:865]: PEP8 line length I imagine

**Comment by lucascolley** [comment]
> If you're happy with this, I'll merge when CI is green?

Let me push a blame-ignore after CI goes green, then yes, thanks!

**Pull request merged**